src/QsCompiler/Transformations/CallGraph.cs

```diff
@@ -54,6 +55,72 @@ internal static bool TryCombineTypeResolutionsForTarget(QsQualifiedName target,
+        /// dictionary to create a lookup whose keys are type parameters and whose values are all the type
+        /// parameters that can be updating by knowing the resolution of the lookup's associated key.
+        /// </summary>
+        private static ILookup<Tuple<QsQualifiedName, NonNullable<string>>, Tuple<QsQualifiedName, NonNullable<string>>> GetReplaceable(TypeParameterResolutions.Builder typeParamResolutions)
+        {
+             return typeParamResolutions
+                .Select(kvp => (kvp.Key, GetTypeParameters.Apply(kvp.Value)))   // Get any type parameters in the resolution type.
+                .SelectMany(tup => tup.Item2.Select(value => (tup.Key, value))) // For each type parameter found, match it to the dictionary key.
+                .ToLookup(                                                      // Reverse the keys and resulting type parameters to make the lookup.
+                    kvp => kvp.value,
+                    kvp => kvp.Key);
+        }
+
+        /// <summary>
+        /// Applies all the type parameter resolutions in a dictionary to the whole dictionary,
+        /// resulting in a resolution dictionary that has type parameter keys that are not
+        /// referenced in the values of the dictionary.
+        /// </summary>
+        internal static bool TryCombineTypeResolutionDictionary(out TypeParameterResolutions combined, TypeParameterResolutions resolutionDictionary)
+        {
+            if (!resolutionDictionary.Any())
+            {
+                combined = TypeParameterResolutions.Empty;
+                return true;
+            }
+
+            var combinedBuilder = ImmutableDictionary.CreateBuilder<Tuple<QsQualifiedName, NonNullable<string>>, ResolvedType>();
+            var success = true;
+
+            static bool IsConstrictiveResolution(Tuple<QsQualifiedName, NonNullable<string>> typeParam, ResolvedType res) =>
+                ConstrictionCheck.Apply(typeParam, res);
+
+            foreach (var (typeParam, paramRes) in resolutionDictionary)
+            {
+                // Contains a lookup of all the keys in the combined resolutions whose value needs to be updated
+                // if a certain type parameter is resolved by the currently processed dictionary.
+                var mayBeReplaced = GetReplaceable(combinedBuilder);
+
+                // Check that we are not constricting a type parameter to another type parameter of the same callable
+                // both before and after updating the current value with the resolutions processed so far.
+                success = success && !IsConstrictiveResolution(typeParam, paramRes);
+                var resolvedParamRes = ResolvedType.ResolveTypeParameters(combinedBuilder.ToImmutable(), paramRes);
+                success = success && !IsConstrictiveResolution(typeParam, resolvedParamRes);
+
+                // Create a dictionary with just the current resolution in it.
```

Comment:
The logic from here to the end of the foreach is the same than the one in a section of `TryCombineTypeResolutions`. It might be worthy to create a helper method that takes `singleResolution` and the `IEnumerable` as parameters.

Reply:
I've added such a method, `UpdatedReplaceableResolutions`, but I'm not sure it was worth it. I needed to add in a lot of parameters to to give it all the context it needs, and it updates the combinedBuilder as a side effect, the return value being the success flag. It just seems a bit awkward. What do you think?

Reply:
I could have passed in `singleResolution` and the result from `mayBeReplaced[typeParam]` as parameters, removing the need for `typeParam` and `mayBeReplaced` as parameters, but that is the same number of parameters and less of the shared logic encapsulated. I don't think that would have made the logic any less complex.

Reply:
I think a bit of the awkwardness can be improved by removing the success flag as a parameter and just return the success that is local to the helper method. The lines in which the helper method is invoked have to be updated to something like the following:
```
success = success && UpdatedReplaceableResolutions(mayBeReplaced, combinedBuilder, typeParam, resolvedParamRes);
```

I think this makes things a bit less awkward since it is a pattern that is already present in other parts of the code.

Having said that, I think the helper method is worth it but I am flexible on whether you want to incorporate it. Maybe we can ask someone else's opinion and go from there?

Reply:
If we don't pass in the success flag, we lose the efficiency gain from not having to traverse the resolved type when checking for constriction when we already know that the validation has failed. This would be simpler if we weren't trying to do the validation at the same time as the combining, but the two tasks are so similar it would be inefficient and needlessly redundant to do them separately.

Reply:
I think that it is better to favor simple and clear code over efficient code, unless the code is demonstrably inefficient (i.e. a profiler or performance test shows that it is too slow to be usable).

Just wanted to make sure I'm understanding the behavior, but

```
success = success && !ConstrictionCheck.Apply(keyInCombined, paramRes);
combinedBuilder[keyInCombined] = ResolvedType.ResolveTypeParameters(singleResolution, combinedBuilder[keyInCombined]);
```

The first line does not always need to run, but the second line should always run? So

```
success = success && UpdatedReplaceableResolutions(mayBeReplaced, combinedBuilder, typeParam, resolvedParamRes);
```

would be incorrect?

Reply:
Yes, but `success` and the method call can be swapped to make it right.

Reply:
I think we could also use the bit-wise `&` to prevent short-circuiting, but that seems a bit esoteric. 😁

Reply:
> 
> 
> I think that it is better to favor simple and clear code over efficient code, unless the code is demonstrably inefficient (i.e. a profiler or performance test shows that it is too slow to be usable).
> 
> Just wanted to make sure I'm understanding the behavior, but
> 
> ```cs
> success = success && !ConstrictionCheck.Apply(keyInCombined, paramRes);
> combinedBuilder[keyInCombined] = ResolvedType.ResolveTypeParameters(singleResolution, combinedBuilder[keyInCombined]);
> ```
> 
> The first line does not always need to run, but the second line should always run? So
> 
> ```cs
> success = success && UpdatedReplaceableResolutions(mayBeReplaced, combinedBuilder, typeParam, resolvedParamRes);
> ```
> 
> would be incorrect?

Yes, it would be incorrect, sorry for the confusion, but as @ScottCarda-MS mentioned, it could be swapped to obtain the correct behavior.

@samarsha what do you think having the helper method? Do you think it is worth having it?

Reply:
I think the helper method is worth having. Talking with @ScottCarda-MS, we had some ideas for simplifying way the success state is threaded through the methods, but no simple solutions for now, so I'm OK with the way it is.